### PR TITLE
Add `git blame`

### PR DESCRIPTION
**Pull request by mrnugget**
This adds a new action to the editor: `editor: toggle git blame`. When used it turns on a sidebar containing `git blame` information for the currently open buffer.

The git blame information is updated when the buffer changes. It handles additions, deletions, modifications, changes to the underlying git data (new commits, changed commits, ...), file saves. It also handles folding and wrapping lines correctly.

When the user hovers over a commit, a tooltip displays information for the commit that introduced the line. If the repository has a remote with the name `origin` configured, then clicking on a blame entry opens the permalink to the commit on the code host.

Users can right-click on a blame entry to get a context menu which allows them to copy the SHA of the commit.

The feature also works on shared projects, e.g. when collaborating a peer can request `git blame` data.

As of this PR, Zed now comes bundled with a `git` binary so that users don't have to have `git` installed locally to use this feature.

### Screenshots

screenshot-2024-03-28-13 57 43@2x

screenshot-2024-03-28-14 01 23@2x
screenshot-2024-03-28-14 01 32@2x

### TODOs

- [x] Bundling `git` binary

### Release Notes

Release Notes:

- Added `editor: toggle git blame` command that toggles a sidebar with git blame information for the current buffer.

### CODE REVIEW

**Comment by d1y** [comment]
The most anticipated enhanced git function is finally coming
I'm ready setting
> Just my opinion😵‍💫

```json
{
	"git": {
	    "blame": {
	      "enabled": true,
	      "format": "<author>, <agoOrDate> • <message>"
	    }
  	}
}
```

and action: `editor::ToggleGitBlame` (default keybind `alt-g-b`)

**Comment by baldwindavid** [comment]
This will be a very handy feature. This is out of scope for now but, while I use blame in VSCode, 99% of the time it's a hover to see which PR introduced the commit and to click the link to view the PR on Github.

File: crates/editor/src/actions.rs [hunk 1]
```diff
@@ -238,13 +238,13 @@ gpui::actions!(
...
         ShowCharacterPalette,
         ShowCompletions,
-        ShowGitBlame,
         ShuffleLines,
         SortLinesCaseInsensitive,
         SortLinesCaseSensitive,
         SplitSelectionIntoLines,
         Tab,
         TabPrev,
+        ToggleGitBlame,
```
Comment (d1y) [inline crates/editor/src/actions.rs:247]: Wow 🎉

**Comment by JosephTLyons** [comment]
Oooooh! This is going to be awesome!

**Comment by mrnugget** [comment]
Current preview:

https://github.com/zed-industries/zed/assets/1185253/d9a65982-2b0f-4254-8f22-37811fd25fb4

**Comment by mrnugget** [comment]
Collaboration works! Now all that's left is bundling `git` and incorporating more design feedback :)

https://github.com/zed-industries/zed/assets/1185253/0db7714c-6253-45ef-ac55-77925ad71d6f

File: crates/editor/src/editor.rs [hunk 2]
```diff
@@ -8824,9 +8840,49 @@ impl Editor {
...
+
+    fn show_git_blame_internal(&mut self, cx: &mut ViewContext<Self>) -> Result<()> {
+        if let Some(project) = self.project.as_ref() {
+            let Some(buffer) = self.buffer().read(cx).as_singleton() else {
+                anyhow::bail!("git blame not available in multi buffers")
+            };
+
+            let project = project.clone();
 
+            // cx.spawn(|this, cx| async move {
```
Comment (mikayla-maki) [inline crates/editor/src/editor.rs:8867]: Probably should delete this

**Comment by mrnugget** [comment]
bundled git binary also works!

screenshot-2024-03-28-18 23 50@2x

**Comment by baldwindavid** [comment]
@mrnugget I've been using this in `main` and it's wonderful! Sorry for the unrelated, probably dumb question, but I've done a `cargo run --release` to get a Zed executable. Are there instructions somewhere to get a Mac "app" like you have there that I can run similarly to Preview/Stable releases?

**Comment by mmtftr** [comment]
Even though this is merged, I have a question @mrnugget. I'm curious as to why the code cannot just use git2::Blame for this. git2::Blame has Blame::blame_buffer which I believe would do the same thing as calling `git blame --contents`. I'm not sure what benefits we get by doing a `--incremental` especially since we wait for the command to exit.

**Comment by mrnugget** [comment]
@baldwindavid yep, that's the one. `scripts/bundle-mac`.

@mmtftr because it's very, very slow. See here: https://github.com/libgit2/libgit2/issues/3027 Running `git blame` on `crates/editor/src/editor.rs` in this repository here took 40s, vs 600ms shelling out to `git`. `--incremental` in this case is only used for its machine-readable output.

**Pull request merged**